File: ts/src/kv/key_value_store.test.ts
import { KeyNotFound, KeyValueStore } from "../types/key_value";
import { SerDe } from "../types/serde";
import { KVInMemory } from "./kv_inmemory";
import { WrapKvStore } from "./wrap_kv";
import { Result } from "wueste/result";

export async function testKV(kv: KeyValueStore<string>) {
  const hiStr = "hi";

  let res = kv.Get("test");
  expect(res).rejects.toEqual(new KeyNotFound("test"));
  res = kv.Set("test", hiStr);
  expect(res).resolves.toEqual(hiStr);
  res = kv.Get("test");
  expect(res).resolves.toEqual(hiStr);
  res = kv.Del("test");
  expect(res).resolves.toEqual(hiStr);
  res = kv.Del("test");
  expect(res).rejects.toEqual(new KeyNotFound("test"));
  res = kv.Get("test");
  expect(res).rejects.toEqual(new KeyNotFound("test"));
}

class Base64 implements SerDe<string> {
  Marshal(t: string): Result<Uint8Array> {
    return Result.Ok(new TextEncoder().encode(t));
  }
  Unmarshal(raw: Uint8Array): Result<string> {
    return Result.Ok(new TextDecoder().decode(raw));
  }
}

it("InMemoryKV", async () => {
  const bkv = new KVInMemory();
  const kv = WrapKvStore<string>(bkv, new Base64());
  await testKV(kv);
});

File: go/kv/key_value_store_test.go
package kv

import (
	"testing"

	"github.com/mabels/runified/kv/kv_inmemory"
	"github.com/mabels/runified/types"
)

func testKV(t *testing.T, kv types.KeyValueStore[string]) {
	hiStr := "hi"
	err := kv.Set("test", &hiStr)
	if err != nil {
		t.Fatal(err)
	}
	v, err := kv.Get("test")
	if err != nil {
		t.Fatal(err)
	}
	if *v != "hi" {
		t.Fatal("value not hi", v)
	}
	prev, err := kv.Del("test")
	if err != nil {
		t.Fatal(err)
	}
	if *prev != "hi" {
		t.Fatal("value not hi", v)
	}
	_, err = kv.Del("test")
	_, isNotFound := err.(types.KeyNotFound)
	if !isNotFound {
		t.Fatal(err)
	}
	_, err = kv.Get("test")
	_, isNotFound = err.(types.KeyNotFound)
	if !isNotFound {
		t.Fatal(err)
	}
}

func TestInMemoryKV(t *testing.T) {
	bkv := kv_inmemory.NewInMemory()
	kv := WrapKvStore[string](bkv, &types.JsonSerDe{}, func() *string { return new(string) })
	testKV(t, kv)
}

func TestFireBaseKV(t *testing.T) {
	// _, err := kv_firebase.NewFirebaseRealtimeDB[kvString](context.Background(), types.FireBaseConfig{
	// 	DBUrl:                 "",
	// 	ServiceAccountKeyPath: "",
	// })
	// if err != nil {
	// 	t.Fatal(err)
	// }
	// testKV(t, kv.(KeyValueStore[kvString]))
}

func TestFireStore(t *testing.T) {
	// _, err := kv_firestore.NewFirestore(context.Background(),
	// 	types.FireBaseConfig{
	// 		DBUrl:                 "",
	// 		ServiceAccountKeyPath: "",
	// 	})
	// if err != nil {
	// 	t.Fatal(err)
	// }
	// testKV(t, kv.(KeyValueStore[kvString]))
}
